File: OdessaMacGUIApp/Source/Workers/ScanWorkerHost.m
//
//  ScanWorkerHost.m
//  OdessaMacGUIApp
//
//  Created by Noah Spitzer-Williams on 8/29/12.
//  Copyright (c) 2012 Authentically Digital LLC. All rights reserved.
//

#import "ScanWorkerHost.h"

#if DEBUG
static const int ddLogLevel = LOG_LEVEL_VERBOSE | LOG_LEVEL_INFO | LOG_LEVEL_ERROR | LOG_LEVEL_WARN;
#else
static const int ddLogLevel = LOG_LEVEL_INFO | LOG_LEVEL_ERROR | LOG_LEVEL_WARN;
#endif


@interface ScanWorkerHost ()

@property unsigned long long inputFilesTotalSizeInKBytes;

@property unsigned long long inputFilesCompletedInKBytes;

@property (strong) NSDate* scanStartTime;

@property (strong) ThumbnailGenerator* thumbnailGenerator;

@property (unsafe_unretained) InputFileObject* activeInputFileObject; // weak because this is a loose reference to it

@property (unsafe_unretained) ScanWorker* activeScanWorker;

@property NSInteger completedScanWorkerCount;

@end


@implementation ScanWorkerHost


- (void)main
{
    DDLogInfo(@"called");
    
    [self.delegate scanWorkerHostStatusUpdated:@"Starting..."];
    
    self.inputFilesTotalSizeInKBytes = [self getInputFilesTotalSize];
    self.inputFilesCompletedInKBytes = 0;
    self.scanStartTime = [NSDate date];
    
    [self.mainModel.highlightObjects removeAllObjects];
    
    self.thumbnailGenerator = [ThumbnailGenerator new];
    
    NSOperationQueue* scanWorkerOperationQueue = [NSOperationQueue new];
    [scanWorkerOperationQueue setMaxConcurrentOperationCount:1];
    
    for (InputFileObject* inputFileObject in self.mainModel.inputFileObjects)
    {
        if (self.isCancelled)
            break;
        
        if (inputFileObject.isSmallDropHereControl)
            continue;
        
        NSString* status = [NSString stringWithFormat:@"Scanning %@...", inputFileObject.sourceURL.lastPathComponent.stringByDeletingPathExtension];
        [self.delegate scanWorkerHostStatusUpdated:status];
        
        ScanWorker* scanWorker = [ScanWorker workerWithInputFileObject:inputFileObject];
        [scanWorker setDelegate:self];
        [scanWorker setMainModel:self.mainModel];
        [scanWorker setThumbnailGenerator:self.thumbnailGenerator];
        
        self.activeInputFileObject = inputFileObject;
        self.activeScanWorker = scanWorker;
        
        [scanWorkerOperationQueue addOperations:@[scanWorker] waitUntilFinished:NO];
        
        while (!scanWorker.isFinished)
        {
            if (self.isCancelled)
                [scanWorker cancelWorker];
            
            [NSThread sleepForTimeInterval:0.5];
        }
        
        
        inputFileObject.scanWorkerResult = scanWorker.scanWorkerResult;
        
        self.completedScanWorkerCount++;
    }
    
    DDLogInfo(@"Highlights found: %ld", self.mainModel.highlightObjects.count);
    
    [self trackAnalytics];
    
    if (self.isCancelled == NO)
    {
        // in case we scan really fast, wait a few seconds so the user sees what happened
        NSTimeInterval minScanTime = 4; // this should be under 5 so we always show "Calculating time remaining..."
        while (-[self.scanStartTime timeIntervalSinceNow] < minScanTime)
            [NSThread sleepForTimeInterval:0.2];
        
        [self.delegate scanWorkerHostFinished];
    }
    else
        [self.delegate scanWorkerHostCancelled];
}


- (void)scanWorkerProgressUpdated:(NSInteger)progress
{
    // doesn't really matter since we use timer
}


- (BOOL)getProgressValue:(NSInteger*)progress
{
    if (self.activeInputFileObject == nil || self.inputFilesTotalSizeInKBytes == 0)
        return NO;
    
    self.inputFilesCompletedInKBytes = 0;
    for (InputFileObject* inputFileObject in self.mainModel.inputFileObjects)
    {
        if (inputFileObject.isSmallDropHereControl)
            continue;
        
        if (inputFileObject.scanWorkerResult == ScanWorkerResultNotFinished ||
            inputFileObject == self.activeInputFileObject)
            continue;
        
        self.inputFilesCompletedInKBytes += inputFileObject.sourceURL.fileSizeInKBytes;
    }
    
    unsigned long long scanWorkersCompletedInKBytes = 0;
    if (self.activeInputFileObject != nil)
    {
        if (self.activeInputFileObject.totalFrames > 0)
        {
  //          DDLogInfo(@"processed frames: %lld", self.activeScanWorker.processedFrames);
            scanWorkersCompletedInKBytes += (unsigned long long)(self.activeScanWorker.processedFrames / (double)self.activeInputFileObject.totalFrames * self.activeInputFileObject.sourceURL.fileSizeInKBytes);
//            DDLogInfo(@"fileSize: %lld", self.activeInputFileObject.sourceURL.fileSize);
            //DDLogVerbose(@"scanWorkersCompletedInKBytes: %lld", scanWorkersCompletedInKBytes);
        }
    }
    
    unsigned long long totalKBytesProcessed = self.inputFilesCompletedInKBytes + scanWorkersCompletedInKBytes;
    
    if (totalKBytesProcessed > 0)
    {
        *progress = (NSInteger)(totalKBytesProcessed * 100 / (double)self.inputFilesTotalSizeInKBytes);
        return YES;
    }
    *progress = 0;
    return NO;
}


- (NSString *)getTimeRemaining
{
    NSTimeInterval elapsedTime = -[self.scanStartTime timeIntervalSinceNow];
    
    if (elapsedTime < 5)
        return @"Calculating time remaining...";
    
    NSInteger currentProgressValue = 0;
    
    if ([self getProgressValue:&currentProgressValue] == false)
        return @"Calculating time remaining...";
    
    if (currentProgressValue > 0)
    {
        NSTimeInterval totalEstimatedTime = elapsedTime * 100 / currentProgressValue;
        
        NSTimeInterval timeRemaining = totalEstimatedTime - elapsedTime;
        
        NSInteger minutes = ((long)timeRemaining / 60) % 60;
        NSInteger hours = timeRemaining / 3600;
        
        if (hours > 0)
            return [NSString stringWithFormat:@"%ldh %ldm left", hours, minutes];
        else if (minutes > 0)
            return [NSString stringWithFormat:@"%ldm left", minutes];
        else
            return @"<1m left";
    }
    
    return @"Calculating time remaining...";
}


- (void)trackAnalytics
{
    
    NSTimeInterval totalScanDuration = -[self.scanStartTime timeIntervalSinceNow];
    if (totalScanDuration > 0)
    {
        NSTimeInterval totalVideoDuration = 0;
        
        NSURL* sampleVideoURL = [MainModel getURLToSampleVideo];
        BOOL isSampleVideoIncluded = NO;
        BOOL areRealVideosIncluded = NO;
        
        for (InputFileObject *inputFileObject in self.mainModel.inputFileObjects)
        {
            if (inputFileObject.scanWorkerResult != ScanWorkerResultSuccess)
                continue; // we only care about good scans
            
            if (inputFileObject.isSmallDropHereControl)
                continue;
            
            if ([inputFileObject.sourceURL.path isEqualToString:sampleVideoURL.path])
                isSampleVideoIncluded = YES;
            else
            {
                areRealVideosIncluded = YES;
            }
            
            totalVideoDuration += inputFileObject.videoDuration;
        }
        
        if (totalVideoDuration > 0 && areRealVideosIncluded)
        {
            [AnalyticsHelper fireEvent:@"Each scan - highlights found" eventValue:@(self.mainModel.highlightObjects.count)];
            
            NSNumber* scanMultiplier = [NSNumber numberWithDouble:totalVideoDuration / totalScanDuration];
            DDLogInfo(@"Average scan multiplier = %@", scanMultiplier);
            
            [AnalyticsHelper fireEvent:@"Each scan - scan multiplier" eventValue:scanMultiplier];
            
            [AnalyticsHelper fireEvent:@"Each scan - video duration in minutes" eventValue:@(floor(totalVideoDuration / 60))];
            
        }
        
        if (isSampleVideoIncluded)
        {
            NSUserDefaults* standardUserDefaults = [NSUserDefaults standardUserDefaults];
            if (standardUserDefaults)
            {
                if ([standardUserDefaults boolForKey:@"HasScannedSampleVideo"] == NO)
                {
                    if ([AnalyticsHelper fireEvent:@"First scan sample video" eventValue:0])
                    {
                        [standardUserDefaults setBool:YES forKey:@"HasScannedSampleVideo"];
                        [standardUserDefaults synchronize];
                    }
                }
            }
        }
        
    }
    

}


- (unsigned long long) getInputFilesTotalSize
{
    DDLogInfo(@"called");
    
    unsigned long long inputFilesSize = 0;
    for (InputFileObject* inputFileObject in self.mainModel.inputFileObjects)
    {
        if (inputFileObject.isSmallDropHereControl)
            continue;
        
        inputFilesSize += inputFileObject.sourceURL.fileSizeInKBytes;
    }
    
    return inputFilesSize;
}

@end

File: OdessaGUIProject/Workers/ScanWorkerHost.cs
using System;
using System.ComponentModel;
using System.Diagnostics;
using System.Threading;
using GaDotNet.Common.Helpers;
using NLog;
using OdessaGUIProject.Workers;
using System.IO;

namespace OdessaGUIProject
{
    internal class ScanWorkerHost : BackgroundWorker
    {
        private static readonly Logger Logger = LogManager.GetCurrentClassLogger();

        /// <summary>
        /// Keeps track of the InputFileObject that's currently being scanned
        /// </summary>
        internal InputFileObject ActiveInputFileObject;

        /// <summary>
        /// How many ScanWorkers have finished. So we can show this in status.
        /// </summary>
        internal int CompletedScanWorkerCount = 0;

        /// <summary>
        /// How many KBytes we've processed
        /// </summary>
        internal long InputFilesCompletedInKBytes;

        /// <summary>
        /// The total file size of all input files
        /// </summary>
        internal long InputFilesTotalSizeInKBytes;

        /// <summary>
        /// The starting time of the scan
        /// </summary>
        internal DateTime ScanStartTime;

        internal ScanWorkerHost()
        {
            DoWork += ScanWorkerHost_DoWork;
            WorkerReportsProgress = true;
            WorkerSupportsCancellation = true;
            IsCancelled = false;
        }

        internal delegate void StatusChangedEventHandler(object sender, DoWorkEventArgs e);

        internal event StatusChangedEventHandler StatusChanged;

        internal bool IsCancelled { get; set; }

        /// <summary>
        /// During the scan, return the progress
        /// </summary>
        /// <returns>True if progress should be updated</returns>
        internal bool GetProgressValue(ref int value)
        {
            if (ActiveInputFileObject == null || InputFilesTotalSizeInKBytes == 0)
                return false;

            InputFilesCompletedInKBytes = 0;
            foreach (var inputFileObject in MainModel.InputFileObjects)
            {
                if (inputFileObject.ScanWorkerResult == ScanWorker.ScanWorkerResults.NotFinished ||
                    inputFileObject == ActiveInputFileObject)
                    continue;

                InputFilesCompletedInKBytes += inputFileObject.SourceFileInfo.Length / 1024;
            }

            long scanWorkersCompletedInKBytes = 0;
            if (ActiveInputFileObject != null)
            {
                if (ActiveInputFileObject.TotalFrames > 0)
                {
                    scanWorkersCompletedInKBytes += (long)
                                                    (ActiveInputFileObject.ScanWorker.ProcessedFrames / (double)ActiveInputFileObject.TotalFrames *
                                                     (ActiveInputFileObject.SourceFileInfo.Length / 1024d));
                    //Console.WriteLine("progress of " + scanWorker.inputFile.Name);
                    //Console.WriteLine(NativeOdessaMethods.GetFramesProcessed() + " of " + NativeOdessaMethods.GetTotalFrames());
                    //Console.WriteLine("duration: " + scanWorker.inputFile.Length / 1024);
                }
                else
                {
                    Console.WriteLine("TotalFrames = 0!");
                }
            }

            //Console.WriteLine("already completed: " + MainModel.InputFilesCompletedInKBytes);
            //Console.WriteLine("this completed: " + ScanWorkersCompletedInKBytes);

            long totalKBytesProcessed = InputFilesCompletedInKBytes + scanWorkersCompletedInKBytes;

            if (totalKBytesProcessed > 0)
            {
                value = (int)(totalKBytesProcessed * 100 / InputFilesTotalSizeInKBytes);
                if (value > 100) // to prevent errors
                {
                    Debug.Assert(true, "Somehow progress value is greater than 100");
                    value = 100;
                }
                //Console.WriteLine("progress: " + value + " (" + TotalKBytesProcessed + " / " + MainModel.InputFilesTotalSizeInKBytes + ")");
                return true;
            }
            value = 0;
            return true;
        }

        /// <summary>
        /// Calculates the time remaining in the scan
        /// </summary>
        /// <returns></returns>
        internal string GetTimeRemaining()
        {
            long elapsedTimeInSeconds = Convert.ToInt64((DateTime.Now - ScanStartTime).TotalSeconds);

            if (elapsedTimeInSeconds < 10)
                // our estimate won't be accurate for several seconds so let's not even display it
                return "Calculating time remaining...";

            int currentProgressValue = 0;

            if (GetProgressValue(ref currentProgressValue) == false)
                return "";

            if (currentProgressValue > 0)
            {
                long totalEstimatedTimeInSeconds = elapsedTimeInSeconds * 100 / currentProgressValue;

                long timeRemainingInSeconds = totalEstimatedTimeInSeconds - elapsedTimeInSeconds;

                var ts = TimeSpan.FromSeconds(timeRemainingInSeconds);

                //logger.Info("Elapsed seconds: " + ElapsedTimeInSeconds + ", TotalKBytesProcessed: " + TotalKBytesProcessed.ToString("0,000,000") +
                //    ", InputFilesTotalSizeInKBytes: " + MainModel.InputFilesTotalSizeInKBytes.ToString("0,000,000"));

                if (ts.Hours > 0)
                    return ts.Hours + "h " + ts.Minutes + "m left";
                if (ts.Minutes > 0)
                    return ts.Minutes + "m left";
                return "<1m left";
            }
            return "Calculating time remaining...";
        }

        /// <summary>
        /// Calculates the total file size of all the input files.
        /// Used to calculate progress.
        /// </summary>
        /// <returns></returns>
        private static long CalculateInputFilesTotalSize()
        {
            long ret = 0;
            foreach (var inputFileObject in MainModel.InputFileObjects)
            {
                ret += (inputFileObject.SourceFileInfo.Length / 1024);
            }
            return ret;
        }

        private void ReportStatusChanged()
        {
            string status;
            if (ActiveInputFileObject != null)
            {
                status = "Scanning " + Path.GetFileNameWithoutExtension(ActiveInputFileObject.SourceFileInfo.Name) + " (" + (CompletedScanWorkerCount + 1) + " of " +
                           MainModel.InputFileObjects.Count + ")";
            }
            else
                status = "Starting...";

            var e = new DoWorkEventArgs(null);
            e.Result = status;
            StatusChanged(this, e);
        }

        private void scanWorker_ProgressChanged(object sender, ProgressChangedEventArgs e)
        {
            // trigger ScanWorkerHost to fire its progressUpdate event
            ReportProgress(0);
        }

        private void ScanWorkerHost_DoWork(object sender, DoWorkEventArgs e)
        {
            // Authorization check happens in ScanWorker

            Logger.Info("ScanWorkerHost started");

            InputFilesTotalSizeInKBytes = CalculateInputFilesTotalSize();
            InputFilesCompletedInKBytes = 0;
            ScanStartTime = DateTime.Now;

            MainModel.HighlightObjects.Clear();

            ActiveInputFileObject = null;


            Logger.Info("DetectionThreshold: " + MainModel.DetectionThreshold.ToString());

            foreach (var inputFileObject in MainModel.InputFileObjects)
            {
                if (CancellationPending)
                    break;

                using (inputFileObject.ScanWorker = new ScanWorker(inputFileObject))
                {
                    inputFileObject.ScanWorker.UseCaptureOffset = MainModel.UseCaptureOffset;
                    ActiveInputFileObject = inputFileObject;

                    inputFileObject.ScanWorker.ProgressChanged += scanWorker_ProgressChanged;

                    ReportStatusChanged();

                    try
                    {
                        inputFileObject.ScanWorker.RunWorkerAsync();
                        while (inputFileObject.ScanWorker.IsBusy)
                        { // wait for the scanworker to finish
                            if (CancellationPending)
                                inputFileObject.ScanWorker.CancelScan();

                            System.Threading.Thread.Sleep(500);
                        }
                        inputFileObject.ScanWorkerResult = inputFileObject.ScanWorker.ScanWorkerResult;
                    }
                    catch (Exception ex)
                    {
                        Logger.Error("Error starting ScanWorker! " + ex.ToString());
                    }
                }

                CompletedScanWorkerCount++;
            }

            ActiveInputFileObject = null;

            Logger.Info("Highlights found: " + MainModel.HighlightObjects.Count);

            TrackAnalytics();

            IsCancelled = CancellationPending; // we use this instead of e.Cancel because if we set e.Cancel, we can't read e.Result (http://bytes.com/topic/c-sharp/answers/519073-asynch-crash-when-e-cancel-set)

            if (!IsCancelled)
            { // in case we scan really fast, wait a few seconds so the user sees what happens
                var minScanTime = TimeSpan.FromSeconds(4);
                while (DateTime.Now.Subtract(ScanStartTime) < minScanTime)
                    Thread.Sleep(200);
            }
        }

        private void TrackAnalytics()
        {

            TimeSpan totalScanDuration = DateTime.Now - ScanStartTime;
            double totalVideoDurationInSeconds = 0;
            if (totalScanDuration.TotalSeconds > 0)
            {
                var sampleVideoPath = MainModel.GetPathToSampleVideo();
                var areRealVideosIncluded = false; 
                var isSampleVideoIncluded = false;

                foreach (var inputFileObject in MainModel.InputFileObjects)
                {
                    var scanWorker = inputFileObject.ScanWorker;
                    if (scanWorker == null)
                        continue;

                    if (scanWorker.ScanWorkerResult != ScanWorker.ScanWorkerResults.Success)
                        continue; // we only care about good scans

                    if (scanWorker.InputFileObject.SourceFileInfo.FullName == sampleVideoPath)
                        isSampleVideoIncluded = true;
                    else
                        areRealVideosIncluded = true;

                    totalVideoDurationInSeconds += scanWorker.InputFileObject.VideoDurationInSeconds;
                }

                if (totalVideoDurationInSeconds > 0 && areRealVideosIncluded)
                {
                    AnalyticsHelper.FireEvent("Each scan - highlights found", MainModel.HighlightObjects.Count);

                    var scanMultiplier = (int)Math.Round(totalVideoDurationInSeconds / totalScanDuration.TotalSeconds, 0);
                    Logger.Info("Average scan multiplier = " + scanMultiplier + "x");

                    AnalyticsHelper.FireEvent("Each scan - scan multiplier", scanMultiplier);

                    AnalyticsHelper.FireEvent("Each scan - video duration in minutes",
                                              (int)Math.Round(totalVideoDurationInSeconds / 60, 0));
                }

                if (isSampleVideoIncluded && Properties.Settings.Default.HasScannedSampleVideo == false)
                {
                    if (AnalyticsHelper.FireEvent("First scan sample video"))
                    {
                        Properties.Settings.Default.HasScannedSampleVideo = true;
                        Properties.Settings.Default.Save();
                    }
                }
            }
        }
    }
}